pkg/clientauth/middleware/cloud-access-policy-token-signer.go

```diff
@@ -0,0 +1,99 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+
+	authlib "github.com/grafana/authlib/authn"
+	sdkhttpclient "github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"k8s.io/apiserver/pkg/endpoints/request"
+
+	infralog "github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type CloudAccessPolicyTokenSignerMiddlewareProvider struct {
+	tokenExchangeClient authlib.TokenExchanger
+}
+
+type cloudAccessPolicyTokenSignerMiddleware struct {
+	tokenExchangeClient authlib.TokenExchanger
+	audiences           []string
+	next                http.RoundTripper
+}
+
+type signerSettings struct {
+	token            string
+	tokenExchangeURL string
+}
+
+var _ http.RoundTripper = &cloudAccessPolicyTokenSignerMiddleware{}
+
+func ProvideCloudAccessPolicyTokenSignerMiddlewareProvider(cfg *setting.Cfg) (*CloudAccessPolicyTokenSignerMiddlewareProvider, error) {
+	clientCfg, err := readSignerSettings(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenExchangeClient, err := authlib.NewTokenExchangeClient(authlib.TokenExchangeConfig{
+		Token:            clientCfg.token,
+		TokenExchangeURL: clientCfg.tokenExchangeURL,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &CloudAccessPolicyTokenSignerMiddlewareProvider{
+		tokenExchangeClient: tokenExchangeClient,
+	}, nil
+}
+
+func (p *CloudAccessPolicyTokenSignerMiddlewareProvider) New(audiences []string) sdkhttpclient.MiddlewareFunc {
+	return func(opts sdkhttpclient.Options, next http.RoundTripper) http.RoundTripper {
+		return &cloudAccessPolicyTokenSignerMiddleware{
+			tokenExchangeClient: p.tokenExchangeClient,
+			audiences:           audiences,
+			next:                next,
+		}
+	}
+}
+
+func (m cloudAccessPolicyTokenSignerMiddleware) RoundTrip(req *http.Request) (res *http.Response, e error) {
+	log := infralog.New("cloud-access-policy-token-signer-middleware")
+
+	namespace, ok := request.NamespaceFrom(req.Context())
+	if !ok {
+		return nil, fmt.Errorf("namespace not found in request context")
+	}
```

Comment:
Testing this out with the [navtree](https://github.com/grafana/grafana/blob/2e3c2df6bf677bde44017112809089aa4284e513/pkg/services/navtree/navtreeimpl/navtree.go#L192) example -- we will want to use the user from the request to get the namespace
```suggestion
	user, err := identity.GetRequester(req.Context())
	if err != nil {
		return nil, err
	}

	namespace := user.GetNamespace()
```